core/protocol/all_services.go

```diff
@@ -257,6 +257,14 @@ func newServices(
 	svcs.epochService.NotifyOnEpoch(svcs.topology.OnEpochEvent, svcs.topology.OnEpochRestore)
 	svcs.epochService.NotifyOnEpoch(stats.OnEpochEvent, stats.OnEpochRestore)
 
+	// The team engine is used to know the team a party belongs to. The computation
+	// of the referral program rewards requires this information. Since the team
+	// switches happen when the end of epoch is reached, it needs to be one of the
+	// last services to register on epoch update, so the computation is made based
+	// on the team the parties belonged to during the epoch and not the new one.
+	svcs.teamsEngine = teams.NewSnapshottedEngine(svcs.epochService, svcs.broker, svcs.timeService)
+	svcs.snapshotEngine.AddProviders(svcs.teamsEngine)
```

Comment:
This changes has to be done in the integration test setup as well.

Reply:
I've done it before, without realising it was borked in the core too.

Reply:
It's just so we have a consistent order of calling the callback between the integraton test and the core.